src/Test/Verilog/SVType.idr

```diff
@@ -42,105 +120,260 @@ import public Test.Common.Utils
-  EqInteger' : EqSVBasic Integer' Integer'
-  EqBit'     : EqSVBasic Bit'     Bit'
-  EqReal'    : EqSVBasic Real'    Real'
-
-data SVType : Type
-data SVArray : SVType -> Nat -> Nat -> Type
-data AllowedInPackedArr : SVType -> Type
+data SVType : Type where
+  -- Implicit : SVType -- Declare an implicit net type
+  RVar : NonIntegerType -> SVType
+  SVar : IntegerScalarType -> SVType
+  VVar : IntegerVectorType -> SVType
+  PackedArr : (t : SVType) -> {p : PABasic t} -> Nat -> Nat -> SVType
+  ||| The main difference between an unpacked array and a packed array is that
+  ||| an unpacked array is not guaranteed to be represented as a contiguous set of bits
+  |||
+  ||| Ashok B. Mehta. Introduction to SystemVerilog, 2021
+  |||
+  ||| 7.4.1
+  ||| Each packed dimension in a packed array declaration shall be specified by a range specification of the form
+  ||| [ constant_expression : constant_expression ]. Each constant expression may be any integer value --
+  ||| positive, negative, or zero, with no unknown (x) or high-impedance (z) bits. The first value may be greater
+  ||| than, equal to, or less than the second value.
+  |||
+  ||| IEEE 1800-2023
+  UnpackedArr : SVType -> Nat -> Nat -> SVType
+  -- Dynamic array
+  -- Associative array
+  -- Queue
+  -- Packed Structure
+  -- Unpacked Structure
+  -- Union
+  -- String
+  -- Chandle
+  -- Enum
+  -- Constant (parameter)
+  -- Class
+  -- Void (?)
+  -- Event
+  -- User-defined
 
-public export
-data SVType = Arr (SVArray t s e) | Var SVBasic
 
 public export
-defaultNetType : SVType
-defaultNetType = Var Wire'
+data SVObject : Type where
+  ||| Syntax 6-2—Syntax for net declarations
+  |||
+  ||| net_declaration16 ::=
+  |||   net_type [ drive_strength | charge_strength ] [ vectored | scalared ]
+  |||     data_type_or_implicit [ delay3 ] list_of_net_decl_assignments ;
+  |||   | nettype_identifier [ delay_control ] list_of_net_decl_assignments ;
+  |||   | interconnect implicit_data_type [ # delay_value ]
+  |||     net_identifier { unpacked_dimension } [ , net_identifier { unpacked_dimension } ] ;
+  ||| net_type ::=
+  |||   supply0 | supply1 | tri | triand | trior | trireg | tri0 | tri1 | uwire | wire | wand | wor
+  ||| drive_strength ::=
+  |||   ( strength0 , strength1 )
+  |||   | ( strength1 , strength0 )
+  |||   | ( strength0 , highz1 )
+  |||   | ( strength1 , highz0 )
+  |||   | ( highz0 , strength1 )
+  |||   | ( highz1 , strength0 )
+  ||| strength0 ::= supply0 | strong0 | pull0 | weak0
+  ||| strength1 ::= supply1 | strong1 | pull1 | weak1
+  ||| charge_strength ::= ( small ) | ( medium ) | ( large )
+  ||| delay3 ::=
+  |||   # delay_value
+  |||   | # ( mintypmax_expression [ , mintypmax_expression [ , mintypmax_expression ] ] )
+  ||| delay_value ::=
+  ||| unsigned_number
+  |||   | real_number
+  |||   | ps_identifier
+  |||   | time_literal
+  |||   | 1step
+  Net : NetType -> (t : SVType) -> {ald : AllowedNetData t} -> SVObject
```

Comment:
Shouldn't `ald` parameter be marked as `auto` (for better readability and by-hand debugging)?

Reply:
Done